TextControl: Convert component to TypeScript

walbo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Converts the `TextControl` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
Refactors the current `TextControl` to TypeScript.

## Testing Instructions
1. Run npm run dev and ensure no typing errors
2. Fire up the Storybook and confirm the `TextControl` continues to function as expected
3. Test the various text controls within the block/site editors in Gutenberg still work

cc: @ciampo

Commit:
TextControl: Covert component to TypeScript

Commit:
Don't export unforwarded component

ciampo commented on packages/components/tsconfig.json:
```diff
@@ -8,7 +8,7 @@
 			"gutenberg-test-env",
 			"jest",
 			"@testing-library/jest-dom",
-			"snapshot-diff",
```
Well spotted, it must be a leftover from my previous PR #39436

ciampo commented on packages/components/src/text-control/types.ts:
```diff
@@ -0,0 +1,34 @@
+import type { BaseControlProps } from '../base-control/types';
+import type { WordPressComponentProps } from '../ui/context';
+
+export type TextControlProps = WordPressComponentProps<
```
We usually wrap props in `WordPressComponentProps` where the component is defined, would you mind following the same convention here too?

<details>
<summary>Example</summary>

```diff
diff --git a/packages/components/src/text-control/index.tsx b/packages/components/src/text-control/index.tsx
index 1ff2325f08..7634e751d1 100644
--- a/packages/components/src/text-control/index.tsx
+++ b/packages/components/src/text-control/index.tsx
@@ -13,10 +13,11 @@ import { forwardRef } from '@wordpress/element';
  * Internal dependencies
  */
 import BaseControl from '../base-control';
+import type { WordPressComponentProps } from '../ui/context';
 import type { TextControlProps } from './types';
 
 function UnforwardedTextControl(
-	props: TextControlProps,
+	props: WordPressComponentProps< TextControlProps, 'input', false >,
 	ref: ForwardedRef< HTMLInputElement >
 ) {
 	const {
diff --git a/packages/components/src/text-control/types.ts b/packages/components/src/text-control/types.ts
index 71f6d822c4..e72efdb611 100644
--- a/packages/components/src/text-control/types.ts
+++ b/packages/components/src/text-control/types.ts
@@ -7,28 +7,23 @@ import type { HTMLInputTypeAttribute } from 'react';
  * Internal dependencies
  */
 import type { BaseControlProps } from '../base-control/types';
-import type { WordPressComponentProps } from '../ui/context';
 
-export type TextControlProps = WordPressComponentProps<
-	Pick<
-		BaseControlProps,
-		'className' | 'hideLabelFromVision' | 'help' | 'label'
-	> & {
-		/**
-		 * A function that receives the value of the input.
-		 */
-		onChange: ( value: string ) => void;
-		/**
-		 * The current value of the input.
-		 */
-		value: string | number;
-		/**
-		 * Type of the input element to render. Defaults to "text".
-		 *
-		 * @default text
-		 */
-		type?: HTMLInputTypeAttribute;
-	},
-	'input',
-	false
->;
+export type TextControlProps = Pick<
+	BaseControlProps,
+	'className' | 'hideLabelFromVision' | 'help' | 'label'
+> & {
+	/**
+	 * A function that receives the value of the input.
+	 */
+	onChange: ( value: string ) => void;
+	/**
+	 * The current value of the input.
+	 */
+	value: string | number;
+	/**
+	 * Type of the input element to render. Defaults to "text".
+	 *
+	 * @default text
+	 */
+	type?: HTMLInputTypeAttribute;
+};
```
</details>

walbo replied:
Will update the PR. Thanks for the heads up.

ciampo commented on packages/components/src/text-control/types.ts:
```diff
@@ -0,0 +1,34 @@
+		/**
+		 * Type of the input element to render. Defaults to "text".
+		 *
+		 * @default text
```
```suggestion
		 * @default 'text'
```

walbo replied:
Thanks.

ciampo commented on packages/components/src/text-control/stories/index.tsx:
```diff
@@ -0,0 +1,61 @@
+	hideLabelFromVision: false,
+	help: 'Help text to explain the input.',
+	type: 'text',
+};
```
I think we'd prefer to have a default story as _vanilla_ as possible. We could add another story that showcases using `help` and `label`. There's no need to specify `hideLabelFromVision` and `type`, since those are already the default prop values.

```suggestion
export const Default: ComponentStory<
	typeof TextControl
> = DefaultTemplate.bind( {} );
Default.args = {};

export const WithLabelAndHelpText: ComponentStory<
	typeof TextControl
> = DefaultTemplate.bind( {} );
WithLabelAndHelpText.args = {
	...Default.args,
	label: 'Label Text',
	help: 'Help text to explain the input.',
};
```

walbo replied:
Sounds good to me.

ciampo commented on packages/components/src/text-control/stories/index.tsx:
```diff
@@ -0,0 +1,61 @@
+			control: { type: null },
+		},
+		value: {
+			control: { type: null },
```
@mirka, what's your stance here? At the moment the `value` control is disabled since the story controls `TextControl`'s value via internal state.

mirka replied:
Sure, I think it's a reasonable pattern to follow for controlled stories.

ciampo commented on packages/components/src/text-control/types.ts:
```diff
@@ -0,0 +1,34 @@
+		/**
+		 * The current value of the input.
+		 */
+		value: string | number;
```
If I were to write this component from scratch, I would type `value` as a `string`.

Since this component is _not_ experimental (and since the purpose of these TypeScript migrations is to introduce the least amount of changes possible), we may want to keep this type as it currently is, but consider updating it in a follow-up (see also this recent related [conversation](https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1112049865))

walbo replied:
I agree it should be `string` in this case. The old jsdocs said `string` but the README file says `string | number`.

ciampo replied:
Let's keep it as `string|number` for this PR. It would be great if we could then open a follow-up PR to change it to be only `string` — we could then assess in that PR if effectively it's a change that can be applied without introducing any breaking change.

Would you be up for that?

walbo replied:
Yes. Lets do that. I will creeate a follow-up PR once this is merged.

ciampo reviewed (commented):
Thank you for helping out on the `@wordpress/components` TypeScript refactor, @walbo !

I just left a couple of inline comments, but this PR is in a great shape already.

Finally, would you mind adding an entry to the pachake's [CHANGELOG](https://github.com/WordPress/gutenberg/blob/1b58a2bfdc6b908c6712c9334365f45bd51ba15d/packages/components/CHANGELOG.md) ? You may want to add an `### Internal` section under `## Unreleased`.

Commit:
Apply suggestions from code review

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

Commit:
Move WordPressComponentProps to the component

Commit:
Merge branch 'trunk' into refactor/text-control-to-typescript

Commit:
Update CHANGELOG.md

mirka commented on packages/components/src/text-control/index.tsx:
```diff
@@ -0,0 +1,84 @@
+import type { WordPressComponentProps } from '../ui/context';
+import type { TextControlProps } from './types';
+
+function UnforwardedTextControl(
```
By the way, sorry about ef3a12e! I fixed up the contributing guide in #40682.

walbo replied:
No worries :) Great that you updated the guide. Did follow the guide and seems to cover most of the stuff you need.

mirka replied:
Oooh, good to hear! 🙌

ciampo replied:
Thank you for the feedback! And of course do let us know if you feel like there's something missing (or something wrong) in the guide 😄

mirka commented on packages/components/src/text-control/stories/index.tsx:
```diff
@@ -0,0 +1,65 @@
+	argTypes: {
+		onChange: {
+			action: 'onChange',
+			control: { type: null },
```
Nit: It looks like we could remove the `control: { type: null }` in this case. Storybook hides the controls for `on*` props by default.

walbo replied:
Thanks. I didn't know that 👍 Removed in [e30a8dd](https://github.com/WordPress/gutenberg/pull/40633/commits/e30a8ddf2e7abf6548bf3b71d0d4df7ddf66f280)

mirka commented on packages/components/src/text-control/stories/index.tsx:
```diff
@@ -0,0 +1,65 @@
+};
+export default meta;
+
+const DefaultTemplate: ComponentStory< typeof TextControl > = ( {
+	...args
+} ) => {
+	const [ value, setValue ] = useState( '' );
+
+	return (
+		<TextControl
+			{ ...args }
+			value={ value }
+			onChange={ ( v ) => {
+				setValue( v );
+			} }
```
Apparently, the Action thing (enabled on [this line](https://github.com/WordPress/gutenberg/pull/40633/files#diff-f162be886476a0f6aaa0dc9958d453e89d1e56ce78a01b1918873dc958d51edeR21)) only works by passing through the `onChange` prop.

```diff
const DefaultTemplate: ComponentStory< typeof TextControl > = ( {
+	onChange,
	...args
} ) => {
	const [ value, setValue ] = useState( '' );

	return (
		<TextControl
			{ ...args }
			value={ value }
			onChange={ ( v ) => {
				setValue( v );
+				onChange( v );
```

https://user-images.githubusercontent.com/555336/165803385-b897a8e2-3ad4-44a8-95f6-091672672c9e.mp4

walbo replied:
Thanks. Added in [e30a8dd](https://github.com/WordPress/gutenberg/pull/40633/commits/e30a8ddf2e7abf6548bf3b71d0d4df7ddf66f280)

mirka reviewed (commented):
Wonderful, thank you for the contribution!

Looks great on my end, I think we're good to go once the Action thing is sorted 👍

Commit:
Update storybook

- Remove control on onChange. Storybook hides on* by default.
- Add onChange to trigger action

mirka reviewed (approved):
Perfect, thank you! 🚀

Commit:
Merge branch 'trunk' into refactor/text-control-to-typescript
